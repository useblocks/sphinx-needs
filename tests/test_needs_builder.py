import json
from pathlib import Path

import pytest
from syrupy.filters import props


@pytest.mark.parametrize("test_app", [{"buildername": "needs", "srcdir": "doc_test/doc_needs_builder"}], indirect=True)
def test_doc_needs_builder(test_app, snapshot):
    app = test_app
    app.build()

    needs_list = json.loads(Path(app.outdir, "needs.json").read_text())
    assert needs_list == snapshot(exclude=props("created"))


@pytest.mark.parametrize(
    "test_app", [{"buildername": "needs", "srcdir": "doc_test/doc_needs_builder_negative_tests"}], indirect=True
)
def test_doc_needs_build_without_needs_file(test_app):
    import os
    import subprocess

    app = test_app

    srcdir = Path(app.srcdir)
    out_dir = os.path.join(srcdir, "_build")

    out = subprocess.run(["sphinx-build", "-b", "needs", srcdir, out_dir], capture_output=True)
    assert not out.stderr
    assert "needs.json found, but will not be used because needs_file not configured." in out.stdout.decode("utf-8")


@pytest.mark.parametrize(
    "test_app", [{"buildername": "html", "srcdir": "doc_test/doc_needs_builder_parallel"}], indirect=True
)
def test_needs_html_and_json(test_app):
    """
    Build html output and needs.json in one sphinx-build
    """
    import json
    import os
    import subprocess

    app = test_app
    app.build()

    needs_json_path = os.path.join(app.outdir, "needs.json")
    assert os.path.exists(needs_json_path)

    srcdir = app.srcdir
    build_dir = os.path.join(app.outdir, "../needs")
    subprocess.run(["sphinx-build", "-b", "needs", srcdir, build_dir], capture_output=True)
    needs_json_path_2 = os.path.join(build_dir, "needs.json")
    assert os.path.exists(needs_json_path_2)

    # Check if the needs.json files from html/parallel build and builder are the same
    with open(needs_json_path) as f1:
        needs_1 = json.load(f1)
        with open(needs_json_path_2) as f2:
            needs_2 = json.load(f2)

            # Just check need-data, as the rest contains not matching timestamps
            need_data_1 = needs_1["versions"]["1.0"]["needs"]
            need_data_2 = needs_2["versions"]["1.0"]["needs"]
            assert need_data_1 == need_data_2
