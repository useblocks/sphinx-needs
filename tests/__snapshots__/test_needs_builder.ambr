# serializer version: 1
# name: test_doc_needs_builder[test_app0]
  dict({
    'current_version': '1.0',
    'project': 'Python',
    'versions': dict({
      '1.0': dict({
        'filters': dict({
        }),
        'filters_amount': 0,
        'needs': dict({
          'TC_001': dict({
            'arch': dict({
            }),
            'avatar': '',
            'closed_at': '',
            'completion': '',
            'constraints': list([
            ]),
            'constraints_passed': True,
            'constraints_results': dict({
            }),
            'content_id': 'TC_001',
            'created_at': '',
            'delete': None,
            'description': '',
            'docname': 'index',
            'doctype': '.rst',
            'duration': '',
            'external_css': 'external_link',
            'external_url': None,
            'full_title': 'Test example',
            'has_dead_links': '',
            'has_forbidden_dead_links': '',
            'hidden': '',
            'id': 'TC_001',
            'id_prefix': '',
            'is_external': False,
            'is_modified': False,
            'is_need': True,
            'is_part': False,
            'jinja_content': None,
            'layout': '',
            'links': list([
            ]),
            'max_amount': '',
            'max_content_lines': '',
            'modifications': 0,
            'params': '',
            'parent_need': '',
            'parent_needs': list([
            ]),
            'parts': dict({
            }),
            'post_template': None,
            'pre_template': None,
            'prefix': '',
            'query': '',
            'section_name': 'TEST DOCUMENT NEEDS Builder',
            'sections': list([
              'TEST DOCUMENT NEEDS Builder',
            ]),
            'service': '',
            'signature': '',
            'specific': '',
            'status': 'open',
            'style': None,
            'tags': list([
            ]),
            'target_id': 'TC_001',
            'template': None,
            'title': 'Test example',
            'type': 'test',
            'type_name': 'Test Case',
            'updated_at': '',
            'url': '',
            'url_postfix': '',
            'user': '',
          }),
          'TC_NEG_001': dict({
            'arch': dict({
            }),
            'avatar': '',
            'closed_at': '',
            'completion': '',
            'constraints': list([
            ]),
            'constraints_passed': True,
            'constraints_results': dict({
            }),
            'content_id': 'TC_NEG_001',
            'created_at': '',
            'delete': None,
            'description': '',
            'docname': 'index',
            'doctype': '.rst',
            'duration': '',
            'external_css': 'external_link',
            'external_url': None,
            'full_title': 'Negative test example',
            'has_dead_links': '',
            'has_forbidden_dead_links': '',
            'hidden': '',
            'id': 'TC_NEG_001',
            'id_prefix': '',
            'is_external': False,
            'is_modified': False,
            'is_need': True,
            'is_part': False,
            'jinja_content': None,
            'layout': '',
            'links': list([
            ]),
            'max_amount': '',
            'max_content_lines': '',
            'modifications': 0,
            'params': '',
            'parent_need': '',
            'parent_needs': list([
            ]),
            'parts': dict({
            }),
            'post_template': None,
            'pre_template': None,
            'prefix': '',
            'query': '',
            'section_name': 'TEST DOCUMENT NEEDS Builder',
            'sections': list([
              'TEST DOCUMENT NEEDS Builder',
            ]),
            'service': '',
            'signature': '',
            'specific': '',
            'status': 'closed',
            'style': None,
            'tags': list([
            ]),
            'target_id': 'TC_NEG_001',
            'template': None,
            'title': 'Negative test example',
            'type': 'test',
            'type_name': 'Test Case',
            'updated_at': '',
            'url': '',
            'url_postfix': '',
            'user': '',
          }),
          'US_63252': dict({
            'arch': dict({
            }),
            'avatar': '',
            'closed_at': '',
            'completion': '',
            'constraints': list([
            ]),
            'constraints_passed': True,
            'constraints_results': dict({
            }),
            'content_id': 'US_63252',
            'created_at': '',
            'delete': None,
            'description': '',
            'docname': 'index',
            'doctype': '.rst',
            'duration': '',
            'external_css': 'external_link',
            'external_url': None,
            'full_title': 'A story',
            'has_dead_links': '',
            'has_forbidden_dead_links': '',
            'hidden': '',
            'id': 'US_63252',
            'id_prefix': '',
            'is_external': False,
            'is_modified': False,
            'is_need': True,
            'is_part': False,
            'jinja_content': None,
            'layout': '',
            'links': list([
            ]),
            'max_amount': '',
            'max_content_lines': '',
            'modifications': 0,
            'params': '',
            'parent_need': '',
            'parent_needs': list([
            ]),
            'parts': dict({
            }),
            'post_template': None,
            'pre_template': None,
            'prefix': '',
            'query': '',
            'section_name': 'TEST DOCUMENT NEEDS Builder',
            'sections': list([
              'TEST DOCUMENT NEEDS Builder',
            ]),
            'service': '',
            'signature': '',
            'specific': '',
            'status': 'in progress',
            'style': None,
            'tags': list([
              '1',
            ]),
            'target_id': 'US_63252',
            'template': None,
            'title': 'A story',
            'type': 'story',
            'type_name': 'User Story',
            'updated_at': '',
            'url': '',
            'url_postfix': '',
            'user': '',
          }),
        }),
        'needs_amount': 3,
      }),
      '2.0': dict({
        'filters': dict({
        }),
        'filters_amount': 0,
        'needs': dict({
          'TEST_01': dict({
            'avatar': '',
            'closed_at': '',
            'completion': '',
            'created_at': '',
            'description': 'TEST_01',
            'docname': 'index',
            'duration': '',
            'external_css': 'external_link',
            'external_url': 'file:///home/daniel/workspace/sphinx/sphinxcontrib-needs/tests/doc_test/external_doc/__error__#TEST_01',
            'full_title': 'TEST_01 DESCRIPTION',
            'hidden': '',
            'id': 'TEST_01',
            'id_complete': 'TEST_01',
            'id_parent': 'TEST_01',
            'id_prefix': '',
            'is_external': True,
            'is_need': True,
            'is_part': False,
            'layout': None,
            'links': list([
              'SPEC_1',
            ]),
            'max_amount': '',
            'max_content_lines': '',
            'parent_need': None,
            'parent_needs': list([
            ]),
            'parent_needs_back': list([
            ]),
            'parts': dict({
            }),
            'post_template': None,
            'pre_template': None,
            'query': '',
            'section_name': '',
            'sections': list([
            ]),
            'service': '',
            'signature': '',
            'specific': '',
            'status': None,
            'style': None,
            'tags': list([
            ]),
            'template': None,
            'title': 'TEST_01 DESCRIPTION',
            'type': 'impl',
            'type_name': 'Implementation',
            'updated_at': '',
            'url': '',
            'user': '',
          }),
          'TEST_02': dict({
            'avatar': '',
            'closed_at': '',
            'completion': '',
            'created_at': '',
            'description': 'TEST_02',
            'docname': 'index',
            'duration': '',
            'external_css': 'external_link',
            'external_url': 'file:///home/daniel/workspace/sphinx/sphinxcontrib-needs/tests/doc_test/external_doc/__error__#TEST_02',
            'full_title': 'TEST_02 DESCRIPTION',
            'hidden': '',
            'id': 'TEST_02',
            'id_complete': 'TEST_02',
            'id_parent': 'TEST_02',
            'id_prefix': '',
            'is_external': True,
            'is_need': True,
            'is_part': False,
            'layout': None,
            'links': list([
              'TEST_01',
              'REQ_1',
            ]),
            'max_amount': '',
            'max_content_lines': '',
            'parent_need': None,
            'parent_needs': list([
            ]),
            'parent_needs_back': list([
            ]),
            'parts': dict({
            }),
            'post_template': None,
            'pre_template': None,
            'query': '',
            'section_name': '',
            'sections': list([
            ]),
            'service': '',
            'signature': '',
            'specific': '',
            'status': 'open',
            'style': None,
            'tags': list([
              'test_02',
              'test',
            ]),
            'template': None,
            'title': 'TEST_02 DESCRIPTION',
            'type': 'req',
            'type_name': 'Requirement',
            'updated_at': '',
            'url': '',
            'user': '',
          }),
          'TEST_03': dict({
            'avatar': '',
            'closed_at': '',
            'completion': '',
            'created_at': '',
            'description': 'AAA',
            'docname': 'subpage_a/subpage_b/subpage',
            'duration': '',
            'external_css': 'external_link',
            'external_url': 'file:///home/daniel/workspace/sphinx/sphinxcontrib-needs/tests/doc_test/external_doc/__error__#TEST_03',
            'full_title': 'AAA',
            'hidden': '',
            'id': 'TEST_03',
            'id_complete': 'TEST_03',
            'id_parent': 'TEST_03',
            'id_prefix': '',
            'is_external': True,
            'is_need': True,
            'is_part': False,
            'layout': None,
            'links': list([
            ]),
            'max_amount': '',
            'max_content_lines': '',
            'parent_need': None,
            'parent_needs': list([
            ]),
            'parent_needs_back': list([
            ]),
            'parts': dict({
            }),
            'post_template': None,
            'pre_template': None,
            'query': '',
            'section_name': '',
            'sections': list([
            ]),
            'service': '',
            'signature': '',
            'specific': '',
            'status': 'open',
            'style': None,
            'tags': list([
            ]),
            'template': None,
            'title': 'AAA',
            'type': 'req',
            'type_name': 'Requirement',
            'updated_at': '',
            'url': '',
            'user': '',
          }),
        }),
        'needs_amount': 6,
      }),
    }),
  })
# ---
